File: server/java/render/src/main/java/com/metaobjects/render/recover/RecoverOptions.java
package com.metaobjects.render.recover;

import java.util.Map;
import java.util.function.Function;

/**
 * Bounded runtime override surface (the "20%"). aliases/normalizers are MERGED with the
 * schema's, runtime winning on key conflict. onField is the single bespoke-coercion hook.
 */
public record RecoverOptions(
        Tolerance tolerance,
        Map<String, String> aliases,
        Map<String, Function<String, Object>> normalizers,
        OnField onField) {

    /** ctx carries the field path and the FieldSpec; return null to fall through to default coercion. */
    @FunctionalInterface
    public interface OnField {
        Object coerce(String fieldPath, String rawValue, FieldSpec spec);
    }

    public RecoverOptions {
        tolerance = tolerance == null ? Tolerance.NORMAL : tolerance;
        aliases = aliases == null ? Map.of() : Map.copyOf(aliases);
        normalizers = normalizers == null ? Map.of() : Map.copyOf(normalizers);
    }

    public static RecoverOptions defaults() {
        return new RecoverOptions(Tolerance.NORMAL, Map.of(), Map.of(), null);
    }

    public RecoverOptions withTolerance(Tolerance t) {
        return new RecoverOptions(t, aliases, normalizers, onField);
    }
}

File: server/csharp/MetaObjects.Render/Recover/RecoverOptions.cs
namespace MetaObjects.Render.Recover;

/// <summary>
/// Bounded runtime override surface (the "20%").
/// <see cref="Aliases"/> and <see cref="Normalizers"/> are MERGED with the schema's,
/// runtime winning on key conflict. <see cref="OnField"/> is the single bespoke-coercion hook.
/// </summary>
public sealed record RecoverOptions(
    Tolerance Tolerance,
    IReadOnlyDictionary<string, string> Aliases,
    IReadOnlyDictionary<string, Func<string, object?>> Normalizers,
    OnField? OnField)
{
    /// <summary>Default options: Normal tolerance, no aliases, no normalizers, no hook.</summary>
    public static RecoverOptions Defaults() =>
        new(
            Tolerance.Normal,
            new Dictionary<string, string>(),
            new Dictionary<string, Func<string, object?>>(),
            null);

    /// <summary>Return a copy with a different tolerance level.</summary>
    public RecoverOptions WithTolerance(Tolerance t) =>
        this with { Tolerance = t };
}
